website/server/libs/groupTasks.js

```diff
@@ -1,24 +1,73 @@
 import * as Tasks from '../models/task';
+import {model as Groups} from '../models/group';
+import {model as Users} from '../models/user/index';
 
 const SHARED_COMPLETION = {
   default: 'recurringCompletion',
   single: 'singleCompletion',
   every: 'allAssignedCompletion',
 };
 
-async function _completeMasterTask (masterTask) {
-  masterTask.completed = true;
+async function _completeOrUncompleteMasterTask (masterTask, completed) {
+  masterTask.completed = completed;
   await masterTask.save();
 }
 
-async function _deleteUnfinishedTasks (groupMemberTask) {
-  await Tasks.Task.deleteMany({
-    'group.taskId': groupMemberTask.group.taskId,
-    $and: [
-      {userId: {$exists: true}},
-      {userId: {$ne: groupMemberTask.userId}},
-    ],
-  }).exec();
+async function _updateAssignedUsersTasks (masterTask, groupMemberTask) {
+  if (groupMemberTask.type == 'todo') {
+    if (groupMemberTask.completed) {
+      // The task was done by one person and is removed from others' lists
+      await Tasks.Task.deleteMany({
+        'group.taskId': groupMemberTask.group.taskId,
+        $and: [
+          {userId: {$exists: true}},
+          {userId: {$ne: groupMemberTask.userId}},
+        ],
+      }).exec();
+    } else {
+      // The task was uncompleted by the group member and should be recreated for assignedUsers
+      let group = await Groups.findById(masterTask.group.id);
+      let userList = [];
+      masterTask.group.assignedUsers.forEach(userId => {
+        let query = {_id: userId};
+        userList.push(query);
+      });
+      // @REVIEW There has to be a better way to do this
+      // async arrow function callback to the find and/or a forEach resulted in Mongoose ParallelSaveErrors
+      let assignedUsers = await Users.find({
+          $or: userList
+        }).exec();
+      for (let i=0; i < assignedUsers.length; i++) {
+        let promises = [];
+        promises.push(group.syncTask(masterTask, assignedUsers[i]));
+        promises.push(group.save());
+        await Promise.all(promises);
+      }
```

Comment:
There are a mix of Promises and async functions in the codebase. Using Promises from the Group schema functions and the callback structure of the Mongoose find() functions was problematic.

Any JS advice here?

Reply:
Hi @MeanderingCode ! Can you link me to where callback functions are used?

The `mongoose.find` function by default uses callbacks but if you call `.exec()` on it at the end a promise will be returned

Reply:
Ah, I realize I've seen `.exec()`s in most places but don't now recall seeing callbacks passed to mongoose.

Reply:
Given that, should I be `await`ing the Promises' resolution, as above and in the `findById()` for the Group on line 29, or using `Promise.then()`?

For the Group `findById()`, and `exec()` seems clean. I don't like the `for` loop on the `assignedUsers`. It seemed to need to execute serially with `await` on the Promises to avoid mongoose `ParallelSaveErrors`, but I see no other uses of `await` in the server side code.

Reply:
Pushed rework. Commenting on it.